core/src/main/java/hudson/model/queue/WorkUnitContext.java

```diff
@@ -130,17 +135,21 @@ public void synchronizeEnd(Queue.Executable executable, Throwable problems, long
      *      the member threads will report {@link InterruptedException}.
      */
     public void synchronizeEnd(Executor e, Queue.Executable executable, Throwable problems, long duration) throws InterruptedException {
-        endLatch.synchronize();
-
-        // the main thread will send a notification
-        WorkUnit wu = e.getCurrentWorkUnit();
-        if (wu.isMainWork()) {
-            if (problems == null) {
-                future.set(executable);
-                e.getOwner().taskCompleted(e, task, duration);
-            } else {
-                future.set(problems);
-                e.getOwner().taskCompletedWithProblems(e, task, duration, problems);
+        // Let code waiting for the build to finish (future.get()) finishes when there is a faulty SubTask by setting 
+        // the future always.
+        try {
+            endLatch.synchronize();
+        } finally {
+            // the main thread will send a notification
+            WorkUnit wu = e.getCurrentWorkUnit();
+            if (wu.isMainWork()) {
+                if (problems == null) {
+                    future.set(executable);
+                    e.getOwner().taskCompleted(e, task, duration);
+                } else {
+                    future.set(problems);
+                    e.getOwner().taskCompletedWithProblems(e, task, duration, problems);
+                }
             }
         }
     }
```

Comment:
I think the real issue is the `abort(Throwable)` method as that could exit also without completing the task, I think something like:

```suggestion
    }


    /**
     * When one of the work unit is aborted, call this method to abort all the other work units.
     */
    public synchronized void abort(Throwable cause) {
        if (cause==null)        throw new IllegalArgumentException();
        if (aborted!=null)      return; // already aborted    
        try {
            aborted = cause;
            startLatch.abort(cause);
            endLatch.abort(cause);

            Thread c = Thread.currentThread();
            for (WorkUnit wu : workUnits) {
                Executor e = wu.getExecutor();
                if (e != null && e != c)
                    e.interrupt();
            }
        } finally {
            if (!future.start.isDone()) {
                future.start.set(cause);
            }
            if (!future.isDone()) {
                future.set(cause);
            }
        }
    }
```